ntoskrnl/ke/bug.c

```diff
@@ -643,28 +645,138 @@ KiDumpParameterImages(IN PCHAR Message,
 
 VOID
 NTAPI
-KiDisplayBlueScreen(IN ULONG MessageId,
-                    IN BOOLEAN IsHardError,
-                    IN PCHAR HardErrCaption OPTIONAL,
-                    IN PCHAR HardErrMessage OPTIONAL,
-                    IN PCHAR Message)
+KiFormatStackBacktrace(IN INT32 Iteration,
+                       OUT CHAR *Buffer,
+                       IN INT32 BufferSize)
 {
-    CHAR AnsiName[75];
+    CHAR AnsiName[64];
+    PLDR_DATA_TABLE_ENTRY LdrEntry;
+    BOOLEAN InSystem;
+    
+    if ((ULONG_PTR)KiStackBacktrace[Iteration].FramePc > (ULONG_PTR)MmHighestUserAddress &&
+        KiPcToFileHeader(KiStackBacktrace[Iteration].FramePc, &LdrEntry, FALSE, &InSystem))
+    {
+        KeBugCheckUnicodeToAnsi(&LdrEntry->BaseDllName,
+                                AnsiName,
+                                sizeof(AnsiName));
+        KiStackBacktrace[Iteration].FramePc = (PVOID)((ULONG_PTR)KiStackBacktrace[Iteration].FramePc - (ULONG_PTR)LdrEntry->DllBase);
 
+        RtlStringCbPrintfA(Buffer,BufferSize, "%2d %-8p %12s:%-8p", Iteration, KiStackBacktrace[Iteration].Frame, AnsiName, KiStackBacktrace[Iteration].FramePc);
+    }
+    else
+        RtlStringCbPrintfA(Buffer, BufferSize, "%2d %-8p %21p", Iteration, KiStackBacktrace[Iteration].Frame, KiStackBacktrace[Iteration].FramePc);
+}
+
+VOID
+NTAPI
+KiDoStackBacktrace()
```

Comment:
This function does not handle garbage or bad frames well.
Why not use `RtlCaptureStackBackTrace` ?